Add memory requests for nodes and controlplane workloads

invidian opened the pull request:
This commit adds memory requests for all nodes and controlplane
workloads. The reason behind it is to better show user available
resources on both worker and controller nodes e.g. when doing 'kubectl
describe node'. This is important while one scales the controlplane
deployments and may prevent node eviction.

The measurment was done on freshly created cluster, with
prometheus-operator and metrics-server deployed,
on controller node and on worker node, so the numbers might be lower
than the numbers on long-running cluster, but they give at least some
initial visibility.

The values were measured using 'systemd-cgtop -m -1 / --depth=1', not
using 'free', as 'systemd-cgtop' also includes page cache usage, the
same way 'kubelet' is measuring the memory usage.

Before the measurment, following command has been executed:
'sync; echo 1 | sudo tee /proc/sys/vm/drop_caches; sleep 10'
To make sure only active memory has been captured.

system.slice uses ~250Mi, init.scope uses ~200Mi, which sums up to
roughly 500Mi, which is needed for system.

Kubelet in /docker slice was using ~100Mi. etcd in /docker slice was
using ~200Mi, so workers has 100Mi reserved for 'kube' and controllers
has 300Mi.

Memory usage for self-hosted components has been measured using the
following command: 'kubectl top pods --sort-by=memory | sort -h -k3 -r'.

Then, the read values were rounded up a bit.

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

invidian committed:
Add memory requests for nodes and controlplane workloads

This commit adds memory requests for all nodes and controlplane
workloads. The reason behind it is to better show user available
resources on both worker and controller nodes e.g. when doing 'kubectl
describe node'. This is important while one scales the controlplane
deployments and may prevent node eviction.

The measurment was done on freshly created cluster, with
prometheus-operator and metrics-server deployed,
on controller node and on worker node, so the numbers might be lower
than the numbers on long-running cluster, but they give at least some
initial visibility.

The values were measured using 'systemd-cgtop -m -1 / --depth=1', not
using 'free', as 'systemd-cgtop' also includes page cache usage, the
same way 'kubelet' is measuring the memory usage.

Before the measurment, following command has been executed:
'sync; echo 1 | sudo tee /proc/sys/vm/drop_caches; sleep 10'
To make sure only active memory has been captured.

system.slice uses ~250Mi, init.scope uses ~200Mi, which sums up to
roughly 500Mi, which is needed for system.

Kubelet in /docker slice was using ~100Mi. etcd in /docker slice was
using ~200Mi, so workers has 100Mi reserved for 'kube' and controllers
has 300Mi.

Memory usage for self-hosted components has been measured using the
following command: 'kubectl top pods --sort-by=memory | sort -h -k3 -r'.

Then, the read values were rounded up a bit.

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

surajssd commented on assets/lokomotive-kubernetes/bootkube/resources/charts/kubernetes/templates/kube-scheduler.yaml:
```diff
@@ -61,3 +61,6 @@ spec:
           timeoutSeconds: 15
+        resources:
+          requests:
+            memory: "30Mi"
```
How much value just adding `requests`(and no `limits`) provides us? Other than just helping in scheduling.

invidian replied:
Helping in scheduling is main task here.

rata commented on assets/lokomotive-kubernetes/aws/flatcar-linux/kubernetes/cl/controller.yaml.tmpl:
```diff
@@ -95,7 +95,8 @@ systemd:
           --register-with-taints=$${NODE_TAINTS} \
-          --volume-plugin-dir=/var/lib/kubelet/volumeplugins
+          --volume-plugin-dir=/var/lib/kubelet/volumeplugins \
+          --kube-reserved=memory=500Mi
```
Where does this value come from?

invidian replied:
The values are a snapshot of `kubectl top pods --sort-by=memory | sort -h -k3 -r` after deploying a cluster from `example` configuration. They should perhaps be configurable, as depending on the cluster size, they will be changing.

Those values are minimal, just to have something in place.

Should I add such comment in commit message?

rata replied:
Yes, please do.

I'd like to get some values from production clusters maybe too... But we can update later, maybe?

rata replied:
Wow, kubelet+docker was using 500mb?

Note `ssh`, etc. should go under `--system-reserved`, say [the docs](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#system-reserved)

invidian replied:
Ah right. I'll use system-reserved then.

invidian replied:
Done. PTAL.

rata commented on assets/lokomotive-kubernetes/bootkube/resources/charts/calico/templates/daemonset.yaml:
```diff
@@ -138,6 +138,7 @@ spec:
           resources:
             requests:
               cpu: 150m
+              memory: "60Mi"
```
Idem

invidian replied:
Will "resolve" as duplicate.

rata reviewed (commented):
I think adding these fields makes sense, but not sure which was the reasoning to pick them. Can you please share the reasoning (and add it to the commit)?

I can't really review if the numbers make sense without knowing how they were calculated :)

invidian commented:
>I think adding these fields makes sense, but not sure which was the reasoning to pick them. Can you please share the reasoning (and add it to the commit)?

>I can't really review if the numbers make sense without knowing how they were calculated :)

Right, sorry for not providing that straight away. I've added methods of measuring etc to the commit message now. Please take a look.

invidian committed:
packet worker: move getting BGP_PEER_ADDRESS inline

To avoid duplicating the template logic and to make the configuration
more readable, as having quote (") right before the template logic is
very confusing.

Signed-off-by: Mateusz Gozdek <mateusz@kinvolk.io>

vbatts commented:
This seems like something that would be done inside slices of cgroups on the host, to ensure the system can still operate. I guess it seems arbitrary and inflexible to declare numbers like this. Right?
